spec/immutable-data-structures.html

```diff
@@ -6,26 +6,288 @@ <h1>Immutable Data Structures</h1>
 
   <emu-clause id="sec-record-objects">
     <h1>Record Objects</h1>
-    <p>TBD</p>
     <emu-clause id="sec-record-constructor">
       <h1>The Record Constructor</h1>
-      <p>TBD</p>
+      <p>The Record constructor:</p>
+      <ul>
+        <li>is the intrinsic object <dfn>%Record%</dfn>.</li>
+        <li>is the initial value of the *"Record"* property of the global object.</li>
+        <li>creates and initializes a new Record object when called as a function.</li>
+        <li>is not intended to be used with the `new` operator or to be subclassed. It may be used as the value of an *extends* clause of a class definition but a *super* call to the Record constructor will cause an exception.</li>
+      </ul>
       <emu-clause id="sec-record-constructor-record-value">
         <h1>Record ( _value_ )</h1>
-        <p>TBD</p>
+        <p>TODO</p>
+      </emu-clause>
+    </emu-clause>
+    <emu-clause id="sec-properties-of-the-record-constructor">
+      <h1>Properties of the Record Constructor</h1>
+      <p>The Record constructor:</p>
+      <ul>
+        <li>has a [[Prototype]] internal slot whose value is %Record.prototype%.</li>
+        <li>has the following properties:</li>
+      </ul>
+      <emu-clause id="sec-record.isrecord">
+        <h1>Record.isRecord ( _arg_ )</h1>
+        <p>The *isRecord* function takes one argument _arg_, and performs the following steps:</p>
+        <emu-alg>
+          1. Return ? IsRecord(_arg_).
+        </emu-alg>
+      </emu-clause>
+      <emu-clause id="sec-record.assign">
+        <h1>Record.assign ()</h1>
+        <p>TODO</p>
+      </emu-clause>
+      <emu-clause id="sec-record.entries">
+        <h1>Record.entries ()</h1>
+        <p>TODO</p>
+      </emu-clause>
+      <emu-clause id="sec-record.fromEntries">
+        <h1>Record.fromEntries ( _iterable_ )</h1>
+        <p>When the *fromEntries* method is called with argument _iterable_, the following steps are taken:</p>
+        <emu-alg>
+          1. Perform ? RequireObjectCoercible(_iterable_).
+          1. TODO: convert _iterable_ to List of Entries?
+        </emu-alg>
+      </emu-clause>
+      <emu-clause id="sec-record.keys">
+        <h1>Record.keys ( _O_ )</h1>
+        <p>When the *keys* function is called with argument _O_, the following steps are taken:</p>
+        <emu-alg>
+          1. Let _obj_ be ? ToObject(_O_)
+          2. Let _nameList_ be ? EnumerableOwnPropertyNames(_obj_, ~key~).
+          3. Return TupleCreate(_nameList_).
+        </emu-alg>
+      </emu-clause>
+      <emu-clause id="sec-record.values">
+        <h1>Record.values ( _O_ )</h1>
+        <p>When the *values* function is called with argument _O_, the following steps are taken:</p>
+        <emu-alg>
+          1. Let _obj_ be ? ToObject(_O_)
+          2. Let _nameList_ be ? EnumerableOwnPropertyNames(_obj_, ~value~).
+          3. Return TupleCreate(_nameList_).

```

Comment:
Probably the TupleCreate operation will make a wrapper object; this operation should be called something different.

Reply:
Ah, yeah that is a good point. Something like `CreateTupleFromList`, like `CreateArrayFromList`?

cc @rricard

Reply:
Sure, works for now; we can always rename later.